### PR TITLE
built-in iterators should be disposable

### LINKED ISSUE #59263: Iterator and AsyncIterator missing Symbol.dispose method in esnext.disposable
### ⚙ Compilation target

ESNext

### ⚙ Library

esnext.disposable

### Missing / Incorrect Definition

Iterator and AsyncIterator are missing types for Symbol.dispose method.

### Sample Code

```TypeScript
function* myIterable() {
  using stack = new DisposableStack();
  stack.defer(console.log('cleanup'));

  yield 1;
  yield 2;
  yield 3;
}

{
  using iter = myIterable();
  //           ^^^^^^^^^^ error
  iter.next();
}
```

```
The initializer of a 'using' declaration must be either an object with a '[Symbol.dispose]()' method, or be 'null' or 'undefined'.
  Property '[Symbol.dispose]' is missing in type 'Generator<number, void, void>' but required in type 'Disposable'.ts(2850)
lib.esnext.disposable.d.ts(34, 5): '[Symbol.dispose]' is declared here.
```

### Documentation Link

The proposal lists these extensions:

- https://github.com/tc39/proposal-explicit-resource-management?tab=readme-ov-file#built-in-disposables

core-js implementation of the proposal includes the new built-ins:

- https://github.com/zloirock/core-js/pull/1151
- https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/esnext.iterator.dispose.js
- https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/esnext.async-iterator.async-dispose.js

### PR DESCRIPTION
**Pull request by rbuckton**
This updates the definitions for `IteratorObject` and `AsyncIteratorObject` to be disposable, per the spec:

- https://arai-a.github.io/ecma262-compare/?pr=3000&id=sec-%25iteratorprototype%25-%40%40dispose
- https://arai-a.github.io/ecma262-compare/?pr=3000&id=sec-%25asynciteratorprototype%25-%40%40asyncdispose

It should be noted that `%IteratorPrototype%` is synonymous with `Iterator.prototype` per the Iterator Helpers proposal and is represented by `IteratorObject` in TypeScript, and that `%AsyncIteratorPrototype%` is synonymous with `AsyncIterator.prototype` per the Async Iterator Helpers proposal and is represented by `AsyncIteratorObject` in TypeScript.

Fixes #59263

### CODE REVIEW

File: src/testRunner/unittests/evaluation/awaitUsingDeclarations.ts [hunk 1]
```diff
@@ -1874,4 +1874,140 @@ describe("unittests:: evaluation:: awaitUsingDeclarations", () => {
             "catch",
         ]);
     });
+
+    it("deterministic collapse of Await", async () => {
+        const { main, output } = evaluator.evaluateTypeScript(
```
Comment (sandersn) [inline src/testRunner/unittests/evaluation/awaitUsingDeclarations.ts:1879]: Are these tests of existing functionality? I don't see how changing the type of [Async]IteratorObject would change its runtime behaviour.

**Review by sandersn — APPROVED** [review]
I like the code change, but I had a couple of questions about the tests.

File: src/testRunner/unittests/evaluation/awaitUsingDeclarations.ts (hunk 1 above)
Comment (rbuckton) [reply src/testRunner/unittests/evaluation/awaitUsingDeclarations.ts:1879]: This test was in the wrong test suite. I merely moved while adding other tests.

File: tests/baselines/reference/usingDeclarationsWithIteratorObject.errors.txt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    
+    function f() {
+        // should pass
+        using it0 = io;
+        using it1 = g;
+        using it2 = Iterator.from(i)
+        using it3 = new MyIterator();
+        using it4 = [].values();
+        using it5 = new Map<string, string>().entries();
+        using it6 = new Set<string>().keys();
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply tests/baselines/reference/usingDeclarationsWithIteratorObject.errors.txt:22]: `it7` is illustrating that the `Iterator` interface (which only describes a minimal Iterator compatible with `for..of`) does not implement `Symbol.dispose`. Only iterators that inherit from the `IteratorObject` interface (which represents the global `Iterator.prototype` value) will have a `Symbol.dispose` by default.

Iterators returned from `values()`/`keys()`/`entries()` on `Map`/`Set`/`Array`/etc. will inherit from `IteratorObject` and thus are disposable.

**Comment by trevorade** [comment]
I'm seeing some build errors that are related to this:

image

Rather than,

```ts
interface IteratorObject<T, TReturn, TNext> extends Disposable {
}
```

Should it have been:

```ts
interface Iterator<T, TReturn, TNext> extends Disposable {
}
```

`IteratorObject` extends `Iterator` so the mismatch of `IteratorObject` being disposable but `Iterator` _not_ being disposable causes compiler errors in other locations.

**Comment by jakebailey** [comment]
No, it being on the interface isn't correct, I think you just need to update your `@types/node` to pick up fixes we made.

**Comment by trevorade** [comment]
Thx for the guidance!

For anyone else looking into this, the referenced PR is:
https://github.com/DefinitelyTyped/DefinitelyTyped/commit/8c36e5365e2c6ace4c00e8f21c7a10e8bbcabac3

**Pull request merged**